### PR TITLE
fix: Run `flutter analyze` for flutter packages

### PR DESCRIPTION
**Pull request by mnayef95**
## Description
Current `melos analyze` command is running `dart analyze` instead of `flutter analyze` on flutter projects
https://github.com/invertase/melos/issues/792

## Type of Change

- [ ] ✨ `feat` -- New feature (non-breaking change which adds functionality)
- [x] 🛠️ `fix` -- Bug fix (non-breaking change which fixes an issue)
- [ ] ❌ `!` -- Breaking change (fix or feature that would cause existing functionality to change)
- [ ] 🧹 `refactor` -- Code refactor
- [ ] ✅ `ci` -- Build configuration change
- [ ] 📝 `docs` -- Documentation
- [ ] 🗑️ `chore` -- Chore

### CODE REVIEW

**Comment by spydon** [comment]
How come you need to run `flutter analyze` instead of `dart analyze`? It should result in the same warnings/errors?

**Comment by mahmuttaskiran** [comment]
> How come you need to run `flutter analyze` instead of `dart analyze`? It should result in the same warnings/errors?

Let me clarify. For Flutter projects using Melos, when neither `melos bootstrap` nor `flutter pub get` is executed for the packages, `dart analyze` will fail. This happens because it does not run `pub get`, and since dependencies are not resolved, `dart analyze` complains about import statements that refer to any unresolved dependencies. (This is just one case; there can be many others, such as referring to a class that belongs to a dependency.)

However, `flutter analyze` ensures that `pub get` is executed before running the analysis, which resolves this issue.

You can reproduce this issue in any Flutter repository using Melos by running `melos clean` and then attempting to analyze the code without bootstrapping the packages.

---
Note: With this fix, do not try to reproduce this issue and understand the problem in the Melos repository itself, as it is not a Flutter workspace and will always use dart analyze, which does not run pub get.

File: packages/melos/test/commands/analyze_test.dart [hunk 1]
```diff
@@ -277,6 +277,34 @@ $ melos analyze
       expect(regex.hasMatch(logger.output.normalizeNewLines()), isTrue);
     });
 
+    test('should run analysis with --concurrency flag', () async {
```
Comment (mahmuttaskiran) [inline packages/melos/test/commands/analyze_test.dart:280]: Update the desc for the test

**Comment by mnayef95** [comment]
Hello @spydon, I hope you are doing well. According to the [Flutter documentation](https://docs.flutter.dev/reference/flutter-cli), we should use `flutter analyze` instead of `dart analyze`. Additionally, as stated by @mahmuttaskiran, `flutter analyze` will also run `pub get`, making it unnecessary to run bootstrap—especially for large projects with over 100 packages.

Screenshot 2024-11-15 at 5 19 32 PM

**Comment by spydon** [comment]
> Hello @spydon, I hope you are doing well. According to the [Flutter documentation](https://docs.flutter.dev/reference/flutter-cli), we should use `flutter analyze` instead of `dart analyze` for flutter projects. Additionally, as stated by @mahmuttaskiran, `flutter analyze` will also run `pub get`, making it unnecessary to run bootstrap—especially for large projects with over 100 packages.
> 
> Screenshot 2024-11-15 at 5 19 32 PM
> 

If that is the case that flutter analyze is recommended over dart analyze for flutter projects (which isn't actually stated in the docs like you say), we should also follow it the other way around. Dart packages should be using `dart analyze`, but in this PR flutter will be used for all pure dart packages too if there is a flutter package present in the repo.

**Comment by spydon** [comment]
> For Flutter projects using Melos, when neither melos bootstrap nor flutter pub get is executed for the packages, dart analyze will fail.

If you don't run `bootstrap` your monorepo won't be properly set up anyways, so interdependencies in the repo will resolve to pub instead of to their local counterparts, so that `dart analyze` doesn't run `pub get` is not concerning imo.

**Comment by mnayef95** [comment]
> Are your packages not using each other in your repository? If they are and you aren't manually setting path dependencies you won't be running static analysis properly on the actual state of the repo, since it will be pulling down released versions instead. 

Yes, the packages are interdependent, and we are using path dependencies. We tried running `flutter analyze` and `dart analyze`. `dart analyze` fails because some URIs can't be resolved unless `pub get` is run first, so we tried `flutter analyze`, and it worked as it will do pub get and that what we need for static analysis CI step.

I've updated the PR to include the check for Dart packages. I don't think it's an optimal solution, but it's what I could do with the current implementation.

**Review by spydon — APPROVED** [review]
Lgtm, thanks for your contribution!

**Review by mahmuttaskiran — APPROVED** [review]

**Pull request merged**